pkg/cluster/cluster.go

```diff
@@ -67,3 +71,111 @@ func (ln *localNode) Observe(ckit.Observer) {
 }
+
+// Clusterer implements the behavior required for operating Flow controllers
+// in a distributed fashion.
+type Clusterer struct {
+	Node Node
+	Mux  *http.ServeMux
+
+	RegisteredComponents map[string]struct{}
+}
+
+// New creates a Clusterer.
+func New(log log.Logger, clusterEnabled bool, addr, joinAddr string) (*Clusterer, error) {
+	// Standalone node.
+	if !clusterEnabled {
+		mux := http.NewServeMux()
+		mux.HandleFunc("/", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			w.Write([]byte("clustering is disabled"))
+			w.WriteHeader(http.StatusBadRequest)
+		}))
+		return &Clusterer{
+			Node: NewLocalNode(addr),
+			Mux:  mux,
+
+			RegisteredComponents: make(map[string]struct{}),
+		}, nil
+	}
+
+	host, portStr, err := net.SplitHostPort(addr)
+	if err != nil {
+		return nil, err
+	}
+	port, err := strconv.Atoi(portStr)
+	if err != nil {
+		return nil, err
+	}
+
+	mux := http.NewServeMux()
+
+	gossipConfig := DefaultGossipConfig
+	gossipConfig.NodeName = uuid.NewString()
+	gossipConfig.AdvertiseAddr = host
+	gossipConfig.ApplyDefaults(port)
+
+	if joinAddr != "" {
+		gossipConfig.JoinPeers = strings.Split(joinAddr, ",")
+	}
+
+	gossipNode, err := NewGossipNode(log, mux, &gossipConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	// Attempt to start the Node by connecting to the peers in gossipConfig.
+	// If we cannot connect to any peers, fall back to bootstrapping a new
+	// cluster by ourselves.
+	err = gossipNode.Start()
+	if err != nil {
+		level.Debug(log).Log("msg", "failed to connect to peers; bootstrapping a new cluster")
+		gossipConfig.JoinPeers = nil
+		err = gossipNode.Start()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Nodes initially join the cluster in the Viewer state. We can move to the
+	// Participant state to signal that we wish to participate in reading or
+	// writing data.
+	err = gossipNode.ChangeState(context.Background(), peer.StateParticipant)
+	if err != nil {
+		return nil, err
+	}
+
+	res := &Clusterer{
+		Node: gossipNode,
+		Mux:  mux,
+
+		RegisteredComponents: make(map[string]struct{}),
+	}
+
+	gossipNode.Observe(ckit.FuncObserver(func(peers []peer.Peer) (reregister bool) {
+		names := make([]string, len(peers))
+		for i, p := range peers {
+			names[i] = p.Name
+		}
+		level.Info(log).Log("msg", "peers changed", "new_peers", strings.Join(names, ","))
+		return true
+	}))
+
+	return res, nil
+}
+
+// Registration starts and stops keeping track of a nodeID so its Update method
+// can be called whenever the cluster state changes.
+func (c *Clusterer) Registration(nodeID string, watch bool) {
+	if watch {
+		c.RegisteredComponents[nodeID] = struct{}{}
+	} else {
+		delete(c.RegisteredComponents, nodeID)
+	}
+}
+
+// IsRegistered reports whether a nodeID is scheduled to have its Update method
+// called whenever the cluster state changes.
+func (c *Clusterer) IsRegistered(nodeID string) bool {
+	_, ok := c.RegisteredComponents[nodeID]
+	return ok
+}
```

Comment:
I'd like for this to not reference components/nodeIDs here. As it is now, this builds an implicit circular dependency with pkg/flow, as pkg/flow directly imports pkg/cluster, and pkg/cluster assumes knowledge about pkg/flow.

Reply:
We don't have to get the API perfect on the first PR, this is much better from the perspective of it not leaking 👍